src/sdk/comment.ts

```diff
@@ -0,0 +1,48 @@
+import { Context } from "./context";
+import { LogReturn } from "@ubiquity-os/ubiquity-os-logger";
+import { sanitizeMetadata } from "./util";
+
+const HEADER_NAME = "Ubiquity";
+
+/**
+ * Posts a comment on a GitHub issue if the issue exists in the context payload, embedding structured metadata to it.
+ */
+export async function postComment(context: Context, message: LogReturn) {
+  if ("issue" in context.payload && context.payload.repository?.owner?.login) {
+    const metadata = createStructuredMetadata(message.metadata?.name, message);
+    await context.octokit.rest.issues.createComment({
+      owner: context.payload.repository.owner.login,
+      repo: context.payload.repository.name,
+      issue_number: context.payload.issue.number,
+      body: [message.logMessage.diff, metadata].join("\n"),
+    });
+  } else {
+    context.logger.info("Cannot post comment because issue is not found in the payload");
+  }
+}
+
+function createStructuredMetadata(className: string | undefined, logReturn: LogReturn) {
+  const logMessage = logReturn.logMessage;
+  const metadata = logReturn.metadata;
+
+  const jsonPretty = sanitizeMetadata(metadata);
+  const stack = logReturn.metadata?.stack;
+  const stackLine = (Array.isArray(stack) ? stack.join("\n") : stack)?.split("\n")[2] ?? "";
+  const caller = stackLine.match(/at (\S+)/)?.[1] ?? "";
+  const ubiquityMetadataHeader = `<!-- ${HEADER_NAME} - ${className} - ${caller} - ${metadata?.revision}`;
+
+  let metadataSerialized: string;
+  const metadataSerializedVisible = ["```json", jsonPretty, "```"].join("\n");
+  const metadataSerializedHidden = [ubiquityMetadataHeader, jsonPretty, "-->"].join("\n");
+
+  if (logMessage?.type === "fatal") {
+    // if the log message is fatal, then we want to show the metadata
+    metadataSerialized = [metadataSerializedVisible, metadataSerializedHidden].join("\n");
```

Comment:
Is this how it should be displayed?

![image](https://github.com/user-attachments/assets/373e7dbb-69d8-4e98-b070-da446400fab0)


![image](https://github.com/user-attachments/assets/8ecd252d-d2b6-4732-a6aa-c6a717b1e149)

Reply:
I think because it allows us to search for the metadata

Reply:
I thought search would be based on the components of the header not including the HTML comment syntax `<!--` at the start of the header. Bit harder to parse I guess without it...

It's very unlikely we'll ever search by `caller` or by `revision` I think, what about you guys?

If we used just the headers we could do something like this `UbiquityOS - <orgName> - <pluginName>`. First item being the name of the bot, just to make the query a bit more specific. If `className` === `LogLevel` then that gives us a type of filter for search if we include it.

Is there a spec or search/parse implementation yet??

Reply:
I can't resolve this but think it can be if double metadata being posted is not a problem